embulk-core/src/main/java/org/embulk/spi/PageBuilder.java

```diff
@@ -270,6 +275,322 @@ public void close()
+     *
+     * |Row| works as a buffer against plugins that may add values incorrectly without |PageBuilder#addRecord|.
+     * It accepts just one value per column while |PageBuilder| can double-store values regardless of columns.
+     * Double-stored values are overwritten.
+     */
+    private static class Row
+    {
+        private static Row newRow(Schema schema)
+        {
+            ColumnValue[] values = new ColumnValue[schema.getColumnCount()];
+            for (Column column : schema.getColumns()) {
+                values[column.getIndex()] = newValue(column);
+            }
+            return new Row(values);
+        }
+
+        private static ColumnValue newValue(Column column)
+        {
+            Type type = column.getType();
+            if (type.equals(Types.BOOLEAN)) {
+                return new BooleanColumnValue(column);
+            }
+            else if (type.equals(Types.DOUBLE)) {
+                return new DoubleColumnValue(column);
+            }
+            else if (type.equals(Types.LONG)) {
+                return new LongColumnValue(column);
+            }
+            else if (type.equals(Types.STRING)) {
+                return new StringColumnValue(column);
+            }
+            else if (type.equals(Types.JSON)) {
+                return new JsonColumnValue(column);
+            }
+            else if (type.equals(Types.TIMESTAMP)) {
+                return new TimestampColumnValue(column);
+            }
+            else {
+                throw new IllegalStateException("Unsupported type " + type.getName());
+            }
+        }
+
+        private final ColumnValue[] values;
+
+        private Row(ColumnValue[] values)
+        {
+            this.values = values;
+        }
+
+        private void setNull(int columnIndex)
+        {
+            values[columnIndex].setNull();
+        }
+
+        private void setBoolean(int columnIndex, boolean value)
+        {
+            values[columnIndex].setBoolean(value);
+        }
+
+        private void setLong(int columnIndex, long value)
+        {
+            values[columnIndex].setLong(value);
+        }
+
+        private void setDouble(int columnIndex, double value)
+        {
+            values[columnIndex].setDouble(value);
+        }
+
+        private void setString(int columnIndex, String value)
+        {
+            values[columnIndex].setString(value);
+        }
+
+        private void setJson(int columnIndex, Value value)
+        {
+            values[columnIndex].setJson(value);
+        }
+
+        private void setTimestamp(int columnIndex, Timestamp value)
+        {
+            values[columnIndex].setTimestamp(value);
+        }
+
+        private void write(PageBuilder pageBuilder)
+        {
+            for (ColumnValue v : values) {
+                v.write(pageBuilder);
+            }
+        }
+    }
+
+    interface ColumnValue
```

Comment:
At the same time with the above comment, nicer to specify modifiers to these `interface` and `class`.